Slashing protection refactor - EIP 3076

mratsim opened the pull request:
This refactors slashing protection.
We call the old format version 0, and this introduces a change in slashing persistent storage, there is no high-level API change.

### Requirements

This requires nim-eth commits:
- https://github.com/status-im/nim-eth/commit/49c40c9b95152fb7bffd36220a02d8638fea9d12
- https://github.com/status-im/nim-eth/pull/315

### List of changes

1. An export CLI tool to investigate potential corruption of the slashing DB on libp2p-small-01.aws-eu-central-1a.nimbus.pyrmont
  and for users to export the Nimbus DB to the interchange format (currently the old v3)
2. We use SQLite as a database instead of a KV-store. This significantly simplifies the code:
  - No need for ad-hoc range queries with high potential of bugs as found in the audit (#1698 #1699).
  - Simplified update, maintenance and audits.
  - Less code (and more comments to justify why)
  - Uses robust, audited and fuzzed SQLite codepath
  - Users can maintain and inspect the DB with standard tooling
  - Pruning is straightforward
  - No O(2n) behavior to check for surrounded and surrounding attestations

The main con is that it is harder to switch away from a DB than from a KV-store. However this is a hypothetical future cost that we can plan for while update, maintenance, audit and user friendliness are benefits realized right away.
Furthermore we only use standard SQL.

More pros/cons directly in the code https://github.com/status-im/nimbus-eth2/blob/b5066d6df400bf12025995dff40dc484b325e003/beacon_chain/validator_slashing_protection.nim#L115-L172

### TODO

- [x] Pass the DB tests
- [x] Update interchange format to EIP3076 (#1990)
  - This requires supporting a "low-watermark" scheme that doesn't store all the metadata on blocks/attestation
    but just slow/epoch threshold
- [x] Integrate the official test vectors
- [ ] Update the CLI conversion tool

### Out-of-scope (for now)

1. A validator can be taken out by changing its local time far in the future and have him attest. Then slashing protection will prevent it as all future votes will be surrounded by this far future attestation leading to inactivity leaks.

- https://ethresear.ch/t/eth2-attack-via-time-servers/8049
- https://github.com/ethereum/eth2.0-specs/pull/2107/files

2. In-memory acceleration structures to avoid reads:
  1. we can probably tune SQLite cache for starters
  2. I suspect the linked lists were one bottleneck when dealing with millions of attestations
      and we need to bench the new DB to see if it is still a bottleneck.
  3. Initial research in the code inspired from collision detection and geospatial data processing: https://github.com/status-im/nimbus-eth2/blob/b5066d6df400bf12025995dff40dc484b325e003/beacon_chain/validator_slashing_protection.nim#L174-L195

### DB schema

In a discussion with @arnetheduck we considered having the same DB schema as Lighthouse as they also use SQLite.
This is possible if their slashing DB is standalone.
The benefit would be that users wouldn't need to run Lighthouse+Nimbus conversion tools to migrate. Also tools would be able to inspect both DB without changes.
We would likely want to introduce a way to track versions if we ever change the DB schema.

mratsim committed:
Create CLI tool for slashing export

mratsim committed:
Use SQLite as a DB instead of a KV-store

mratsim committed:
Update slashing protection to use the DB

zah commented on beacon_chain/validator_slashing_protection.nim:
```diff
@@ -168,185 +257,270 @@ type
+    WHERE
+      validator_id = ?
+      AND
+      source_epoch < ? AND ? < target_epoch
```
Looking at all the queries, It looks like you'll need to create an index for the `source_epoch` and `target_epoch` columns:

https://sqlite.org/lang_createindex.html

mratsim committed:
Fix init

mratsim committed:
Keep v1 and v2 DBs around

mratsim committed:
Common slashing types

mratsim committed:
DB query multiversioning [skip CI] - requires https://github.com/status-im/nim-stew/pull/71

mratsim committed:
multiversioning of update procs

mratsim committed:
Implement DB multiversioning + slashing IR for interchange serialization and version migration

mratsim committed:
update nim-stew, fix rebase

mratsim committed:
Make stuff compile after rebase. Once again the generic sandwich rears it's ugly head

https://github.com/nim-lang/Nim/issues/11225

mratsim committed:
Fix the interchange sanity check

mratsim committed:
Activate DB updates multiversioning, pass slashing protection tests

mratsim committed:
Use the same schema as Lighthouse v1.1.0

mratsim committed:
Add low-watermark for block (but part of the test suite is now contradicting spec)

mratsim committed:
Low Watermark for attestations

mratsim committed:
genesis_validatorS_root typo

mratsim committed:
Test runner for slashing interchange (need partially valid import support) [skip CI]

mratsim committed:
bump eth2-scenarios

mratsim committed:
Implement partial imports

mratsim committed:
low watermark condition 4 is strict inequality

mratsim committed:
allow default initialized signing root and check the proper "should_succeed" field

mratsim committed:
Sort import to avoid min slashing violations

mratsim committed:
Support duplicate blocks on import

mratsim committed:
Don't consider duplicate imports as a failure

mratsim committed:
Filter out NULL result from aggregate of empty tables

mratsim committed:
More logs

mratsim committed:
Pass all interchange tests + skeleton of finalization pruning

mratsim committed:
improve comments

mratsim committed:
Remove tests that would violate v5 / minimal slashing DB and MinSlot rules

mratsim committed:
With min epoch violations and surround voting, backtracking append for attestations don't make sense anymore

mratsim committed:
Migration tool

mratsim commented:
The PR is in ready for review.

It includes:
- EIP-3076 complete support, with all test vectors.
- An "intermediate representation" called SPDIR (Slashing Protection Database Intermediate Representation) was introduced.
  - DBv1 and DBv2 supports imports and exports to SPDIR
  - SPDIR can be loaded or dumped to SPDIF (Interchange Format)
  - Migration between DBv1 and DBv2 goes through SPDIR
  - If DB schema needs to evolve this allows us to support multiple versions without too much issues.
- We embrace SQLite as a backend which gives us proper support for range queries
  and easier maintenance and code readability
  - The DB schema is modeled after Lighthouse. After discussion with them, there is a benefit for devs to be able to use the same scripts to inspect the DB and also to be able to compare IO overhead as all clients are struggling with slashing IO overhead. However for end users we don't want them to bypass the strict checks mandated by EIP-3076 as that would put validator at risk so we shouldn't advertise copy-pasting between Nimbus<-> Lighthouse.
  - All deletion scripts are protected by a guard that would prevent deleting the most recent block slot, att source_epoch, att target_epoch to prevent mistakes that would wipe the slashing DB.
  - EIP-3076 last test `multiple_interchanges_single_validator_single_message_gap` mandates support for pruning and the minimal config so that interchange files can be included in arbitrary order instead of from the earliest to the latest one. Concretely, after each interchange import you need to prune the DB to only keep the the latest entry otherwise the `MinSlotViolation` checks are polluted by early entries.
- Migration from the old DB to the new one is included at startup. The old DB is kept around but unused.
  - The old DB could be dropped in the future, I'm unsure if there is any overhead of unused tables apart from space.
- The `ncli_slashing` CLI tool has been updated to dump the DB to the new format.
- pruning per validator (and keep only the most recent block, attestations)

What is not included at the moment:
- pruning after finalized epoch
- `ncli_slashing` CLI command for importing from the interchange format

mratsim commented:
Mysterious type error on 32-bit
![image](https://user-images.githubusercontent.com/22738317/106893052-47a1cb00-66ed-11eb-9a3f-0a138bcc6d5b.png)

mratsim commented:
Don't merge yet, trying on my Pyrmont validator.

mratsim commented:
On my beefy machine migration takes 3 min for 1 validator
![image](https://user-images.githubusercontent.com/22738317/106922251-27820400-670d-11eb-965f-a3443cb3ff47.png)

unfortunately this is way too slow so I will use an alternative scheme that import just the low-watermark of the v1 DB since those are actually the first nodes of the linked list.

mratsim committed:
Properly query if a table exist

mratsim committed:
Use low-watermark scheme for faster migration of large number of validators

mratsim commented:
Migration from v1 to v2 should be very fast now as I switched to the low-watermark scheme. There are 2 queries per validators instead of `2 + numBlocks + numAttestations`.

mratsim committed:
update allTests.md

zah commented on beacon_chain/validator_protection/slashing_protection.nim:
```diff
@@ -0,0 +1,441 @@
+       source: Epoch,
+       target: Epoch
+     ): Result[void, BadVote] =
+  ## Returns an error if the specified validator
```
This comment is copy pasted from the function above? Is also correct for this function which checks for attestations?

mratsim replied:
No it's wrong good catch

zah commented on beacon_chain/validator_protection/slashing_protection_common.nim:
```diff
@@ -0,0 +1,443 @@
+
+proc readValue*(reader: var JsonReader, value: var PubKey0x)
+               {.raises: [SerializationError, IOError, ValueError, Defect].} =
+  value = PubKey0x reader.readValue(string).hexToByteArray[:RawPubKeySize]()
```
The serialization library will try to enforce that only `SerializationError` and `IOError` can escape through the high-level API. For this reason, you must catch `ValueError` here and translate to `SerializationError` with `raiseUnexpectedValue` as done few lines below.

zah commented on beacon_chain/validator_protection/slashing_protection_common.nim:
```diff
@@ -0,0 +1,443 @@
+
+proc readValue*(r: var JsonReader, a: var (SlotString or EpochString))
+               {.raises: [SerializationError, IOError, ValueError, Defect].} =
+  a = (typeof a)(r.readValue(string).parseBiggestUint())
```
Ditto (`ValueError`).

zah commented on beacon_chain/validator_protection/slashing_protection_v2.nim:
```diff
@@ -0,0 +1,1167 @@
+      INSERT INTO
+        metadata(slashing_db_version, genesis_validators_root)
+      VALUES
+        (""" & $db.typeof().version() & """, ?);
```
Since the version is an `INTEGER` column, it shouldn't be necessary to quote it as a string?

mratsim replied:
It's a string literal here though, the query has no parameter.

zah replied:
Ah, silly me. Got confused by the triple quotes literal.

zah commented on beacon_chain/validator_protection/slashing_protection_v2.nim:
```diff
@@ -0,0 +1,1167 @@
+    db.backend.exec("""
+      CREATE TABLE signed_attestations(
+          validator_id INTEGER NOT NULL,
+          source_epoch INTEGER NOT NULL,
```
With your latest queries, you can still benefit from a multi-column index on the columns `(validator_id, source_epoch, target_epoch)`.

https://sqlite.org/lang_createindex.html

mratsim replied:
(validator_id, target_epoch) is tagged unique so there should be an index (?), also if we do pruning every finalized epoch which is necessary according to Lighthouse and Prysm experience, we would have a rolling 64 attestations per validator in the DB 2x32 epochs in the general case, I think 64 linear compare vs 6 logarithmic compare aren't that much different, especially if the DB is small enough to always be in memory. An index would make the DB much larger.

This is likely to early to decide before we enable pruning and measure IOPS.

zah commented on beacon_chain/validator_protection/slashing_protection_v2.nim:
```diff
@@ -0,0 +1,1167 @@
+      signing_root
+    FROM
+      signed_attestations
+    WHERE 1=1
```
Why do we have `WHERE 1=1`? This technique is usually used in scripting languages where you append the where clauses conditionally, which is not the case here.

mratsim replied:
It's my SQL script style, I use it so that all lines start with a `and` constraint instead of having a lone wolf.
It's especially useful to easily toggle the first constraint on/off without removing/adding a `and` on the other lines.

zah commented on beacon_chain/validator_protection/slashing_protection_v2.nim:
```diff
@@ -0,0 +1,1167 @@
+  ##       this doesn't check the genesis validator root
+  ##
+  ## Privacy: This leaks user folder hierarchy in case the file does not exist
+  let path = basepath/dbname&".sqlite3"
```
@stefantalpalaru has been changing all usages of the `/` operator due to cross-compilation concerns. Perhaps he would want to change it here as well.

stefantalpalaru replied:
I need a literal forward slash character when the path is used at compile time. Runtime is not a problem.

mratsim commented on beacon_chain/validator_protection/slashing_protection.nim:
```diff
@@ -0,0 +1,441 @@
+        # fatal "Slashing protection DB has an internal error",
+        #   query = queryStr,
+        #   dbV1_result = res1,
+        #   dbV2_result = res2
```
Anyway to allow uncommenting the logs @zah ? I get

```
/[...]/nim-beacon-chain/vendor/nim-chronicles/chronicles/scope_helpers.nim(18, 18) logIMPL
/[...]/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/core/macros.nim(1279, 5) $
/[...]/nim-beacon-chain/beacon_chain/validator_protection/slashing_protection.nim(266, 5) template/generic instantiation of `queryVersions` from here
/[...]/nim-beacon-chain/beacon_chain/validator_protection/slashing_protection.nim(232, 15) template/generic instantiation of `fatal` from here
/[...]/nim-beacon-chain/vendor/nim-chronicles/chronicles.nim(363, 10) template/generic instantiation of `log` from here
/[...]/nim-beacon-chain/beacon_chain/validator_protection/slashing_protection.nim(267, 32) Error: Invalid node kind nnkCall for macros.`$`
```

in chronicles. I think it doesn't work with `astToStr`

zah replied:
Addressed in 
6cc3044

mratsim committed:
Fix doc [skip CI]

mratsim committed:
catch ValueErrors: https://github.com/status-im/nimbus-eth2/pull/2094#discussion_r571966909

zah committed:
Fix a Chronicles compilation error

The `query` parameter of the template was also used as a log statement
property. This forced Nim to replace the property name with the input
expression of the template which resutled in the hard to decipher error.

mratsim commented:
Just so that we are clear on the default and what will happen once this is released.

## Migration

On finding an v1 DB and no v2 DB, old data will be **copied** in the new table. The old data will not be pruned and will stay in the dabase. This will allow rollback if necessary, though if we have data in both DB the old data will be stale.

## Default

The default is to use the new v2 scheme, https://github.com/status-im/nimbus-eth2/blob/6cc304400273e75ea76dde58b7ff62281bca6f1f/beacon_chain/validator_protection/slashing_protection.nim#L143-L157

For a couple of reasons, we can run in dual validation mode on some of our machines but when released to end users we need to not bottleneck on IO.

Also despite being newer untested code and unaudited code, the V2 code is easier to maintain, used fuzzed SQLite codepath, is faster and is the only one properly supporting slashing interchange import (no low watermark in V1) which means it can use all the tests published in the interop repo.
